File: front/public/game/localApi/Vec2.js
class Vec2 {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }
  static create(x, y) {
    return new Vec2(x, y);
  }

  clone() {
    return new Vec2(this.x, this.y);
  }

  copy(b) {
    this.x = b.x;
    this.y = b.y;
    return this;
  }

  add(b) {
    this.x += b.x;
    this.y += b.y;
    return this;
  }
  static add(a, b) {
    return a.clone().add(b);
  }

  sub(b) {
    this.x -= b.x;
    this.y -= b.y;
    return this;
  }
  static sub(a, b) {
    return a.clone().sub(b);
  }

  mul(b) {
    this.x *= b.x;
    this.y *= b.y;
    return this;
  }
  static mul(a, b) {
    return a.clone().mul(b);
  }

  div(b) {
    if (b.x === 0 || b.y === 0) {
      throw new Error('Division by zero');
    }
    this.x /= b.x;
    this.y /= b.y;
    return this;
  }
  static div(a, b) {
    return a.clone().div(b);
  }

  dot(b) {
    return this.x * b.x + this.y * b.y;
  }
  static dot(a, b) {
    return a.dot(b);
  }

  cross(b) {
    return this.x * b.y - this.y * b.x;
  }
  static cross(a, b) {
    return a.cross(b);
  }

  scale(s) {
    this.x *= s;
    this.y *= s;
    return this;
  }
  static scale(a, s) {
    return a.clone().scale(s);
  }

  normalize() {
    const len = this.length();
    if (len !== 0) {
      this.scale(1 / len);
    }
    return this;
  }
  static normalize(a) {
    return a.clone().normalize();
  }

  lengthSquared() {
    return this.x * this.x + this.y * this.y;
  }
  static lengthSquared(a) {
    return a.lengthSquared();
  }

  length() {
    return Math.sqrt(this.lengthSquared());
  }
  static length(a) {
    return a.length();
  }

  static distanceSquared(a, b) {
    return Vec2.sub(b, a).lengthSquared();
  }

  static distance(a, b) {
    return Math.sqrt(Vec2.distanceSquared(a, b));
  }

  angle() {
    return Math.atan2(this.y, this.x);
  }
  static angle(a) {
    return a.angle();
  }

  reflect(normal) {
    const d = this.dot(normal);
    this.sub(Vec2.scale(normal, 2 * d));
    return this;
  }
  static reflect(a, normal) {
    return a.clone().reflect(normal);
  }

  rotate(angle) {
    const cos = Math.cos(angle);
    const sin = Math.sin(angle);
    const x = this.x * cos - this.y * sin;
    const y = this.x * sin + this.y * cos;
    this.x = x;
    this.y = y;
    return this;
  }
  static rotate(a, angle) {
    return a.clone().rotate(angle);
  }
}

export default Vec2;

File: game/app/api/engine/Vec2.py
import math

class Vec2:
  def __init__(self, x, y):
    self.x = x
    self.y = y

  def json(self):
    return {
      'x': self.x,
      'y': self.y,
    }

  def clone(self):
    return Vec2(self.x, self.y)

  def copy(self, b):
    self.x = b.x
    self.y = b.y
    return self

  def add(self, b):
    self.x += b.x
    self.y += b.y
    return self

  @staticmethod
  def static_add(a, b):
    return a.clone().add(b)

  def sub(self, b):
    self.x -= b.x
    self.y -= b.y
    return self

  @staticmethod
  def static_sub(a, b):
    return a.clone().sub(b)

  def mul(self, b):
    self.x *= b.x
    self.y *= b.y
    return self

  @staticmethod
  def static_mul(a, b):
    return a.clone().mul(b)

  def div(self, b):
    if b.x == 0 or b.y == 0:
      raise ValueError('Division by zero')
    self.x /= b.x
    self.y /= b.y
    return self

  @staticmethod
  def static_div(a, b):
    return a.clone().div(b)

  def dot(self, b):
    return self.x * b.x + self.y * b.y

  @staticmethod
  def static_dot(a, b):
    return a.dot(b)

  def cross(self, b):
    return self.x * b.y - self.y * b.x

  @staticmethod
  def static_cross(a, b):
    return a.cross(b)

  def scale(self, s):
    self.x *= s
    self.y *= s
    return self

  @staticmethod
  def static_scale(a, s):
    return a.clone().scale(s)

  def normalize(self):
    len = self.length()
    if len != 0:
      self.scale(1 / len)
    return self

  @staticmethod
  def static_normalize(a):
    return a.clone().normalize()

  def length_squared(self):
    return self.x ** 2 + self.y ** 2

  @staticmethod
  def static_length_squared(a):
    return a.length_squared()

  def length(self):
    return math.sqrt(self.length_squared())

  @staticmethod
  def static_length(a):
    return a.length()

  @staticmethod
  def static_distance_squared(a, b):
    return Vec2.sub(b, a).length_squared()

  @staticmethod
  def static_distance(a, b):
    return math.sqrt(Vec2.distance_squared(a, b))

  def angle(self):
    return math.atan2(self.y, self.x)

  @staticmethod
  def static_angle(a):
    return a.angle()

  def reflect(self, normal):
    d = self.dot(normal)
    self.sub(Vec2.scale(normal, 2 * d))
    return self

  @staticmethod
  def static_reflect(a, normal):
    return a.clone().reflect(normal)

  def rotate(self, angle):
    cos = math.cos(angle)
    sin = math.sin(angle)
    x = self.x * cos - self.y * sin
    y = self.x * sin + self.y * cos
    self.x = x
    self.y = y
    return self

  @staticmethod
  def static_rotate(a, angle):
    return a.clone().rotate(angle)
